include/nbl/video/IGPUDescriptorSet.h

```diff
@@ -28,10 +28,34 @@ class NBL_API IGPUDescriptorSet : public asset::IDescriptorSet<const IGPUDescrip
 		using base_t = asset::IDescriptorSet<const IGPUDescriptorSetLayout>;
 
 	public:
-		IGPUDescriptorSet(core::smart_refctd_ptr<const ILogicalDevice>&& dev, core::smart_refctd_ptr<const IGPUDescriptorSetLayout>&& _layout) : base_t(std::move(_layout)), IBackendObject(std::move(dev)) {}
+		IGPUDescriptorSet(core::smart_refctd_ptr<const ILogicalDevice>&& dev, core::smart_refctd_ptr<const IGPUDescriptorSetLayout>&& _layout, core::smart_refctd_ptr<IDescriptorPool>&& pool, const uint32_t* descriptorStorageOffsets)
+			: base_t(std::move(_layout)), IBackendObject(std::move(dev)), m_pool(std::move(pool))
+		{
+			memcpy(m_descriptorStorageOffsets, descriptorStorageOffsets, asset::EDT_COUNT * sizeof(uint32_t));
+		}
+
+		uint8_t* getDescriptorMemory(const asset::E_DESCRIPTOR_TYPE type, const uint32_t binding) const;
 
 	protected:
-		virtual ~IGPUDescriptorSet() = default;
+		virtual ~IGPUDescriptorSet()
+		{
+			for (const auto& b : getLayout()->getBindings())
+			{
+				assert(m_descriptorStorageOffsets[b.type] != ~0u && "Descriptor of this type doesn't exist in the set!");
+
+				auto* descriptorMemory = getDescriptorMemory(b.type, b.binding);
+				assert(descriptorMemory);
+
+				auto* descriptors = reinterpret_cast<core::smart_refctd_ptr<const asset::IDescriptor>*>(descriptorMemory);
+
+				for (auto i = 0; i < b.count; ++i)
+					descriptors[i].~smart_refctd_ptr();
```

Comment:
this is going to blow up if your descriptor set does not default construct the info structs first (so they're primed with `nullptr`)

This is why I was talking about Dependency Injection

```cpp
class IDescriptorSet
{
   protected:
      ~IDescriptorSet()
      {
         std::destroy_n(getSamplerRefcountingStorage(),m_layout->getMutableSamplerCount());
         for (auto type=0u; type<asset::EDT_COUNT; type++)
            std::destroy_n(getDescriptorRefcountingStorage(type),m_layout->getTotalDescriptorCount(type));
      }
      // gets the allocated storage
      virtual core::smart_refctd_ptr<IDescriptor>* getDescriptorRefcountingStorage(const E_DESCRIPTOR_TYPE type) = 0;
      virtual core::smart_refctd_ptr<ISampler>* getSamplerRefcountingStorage() = 0;
      // gets called by derived class constructor as soon as setup necessary for `get***RefcountringStorage` is complete 
      inline void createDescriptors()
      {
         allocateDescriptors();
         std::uninitialized_default_construct_n(getSamplerRefcountingStorage(),m_layout->getMutableSamplerCount());
         for (auto type=0u; type<asset::EDT_COUNT; type++)
            std::uninitialized_default_construct_n(getDescriptorRefcountingStorage(type),m_layout->getTotalDescriptorCount(type));
      }
      
      uint32_t m_mutableSamplerStorageOffset;
      uint32_t m_descriptorStorageOffsets[asset::EDT_COUNT];
};
```

Reply:
Now we default-construct with `std::default_construct_n` in the `IGPUDescriptorSet` ctor.
Resolved.